### PR TITLE
Set child orders to be children of current order parent before deleting.

### LINKED ISSUE #35909: [HPOS] Fatal error in order list table when parent order is permanently deleted.
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

During testing one of the deposits [PR](https://github.com/woocommerce/woocommerce-deposits/pull/480), I came across this bug. WooCommerce throws a fatal error on the order list table when child orders are there and parent order is permanently deleted. This is happen with HPOS enabled only. 

This is happening because we are trying to set the parent id of the deleted order on child orders. I think we should add a condition here to check if the parent order exists before we set the `parent_id` prop of child orders.
```
PHP Fatal error:  Uncaught WC_Data_Exception: Invalid parent ID in /Users/******/Local Sites/hpos/app/public/wp-content/plugins/woocommerce/includes/abstracts/abstract-wc-data.php:880
Stack trace:
#0 /Users/******/Local Sites/hpos/app/public/wp-content/plugins/woocommerce/includes/abstracts/abstract-wc-order.php(574): WC_Data->error('order_invalid_p...', 'Invalid parent ...')
#1 /Users/******/Local Sites/hpos/app/public/wp-content/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php(1323): WC_Abstract_Order->set_parent_id('150')
#2 /Users/******/Local Sites/hpos/app/public/wp-content/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php(1306): Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore->set_order_prop(Object(WC_Order), 'parent_id', '150')
#3 /Users/******/Local Sites/hpos/app/public/wp-content/plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php(1035): Automattic\WooCommerce\Internal\DataStores\Orders\Orders in /Users/******/Local Sites/hpos/app/public/wp-content/plugins/woocommerce/includes/abstracts/abstract-wc-data.php on line 880
```

image

### Expected behavior

No fatal error should be there, child orders should be listed without any errors, as it displays currently for orders with the WP_Post table. 

### Actual behavior

A fatal error is thrown on the order list table.

### Steps to reproduce

1. Enable HPOS
2. Install and Activate "WooCommerce Deposits" plugin
3. Create a Payment Plan.
4. Create a Product, and enable payment plan deposit on it.
5. Add created a product to the cart with deposit options, and place an order.
6. You will see some scheduled orders are also created (child orders)
7. Move created order and scheduled order to the trash.
8. Permanently delete only created order(parent order).
9. You will see the fatal error on the trash order list.

### WordPress Environment

<details>

`
### WordPress Environment ###

WordPress address (URL): https://hpos.local
Site address (URL): https://hpos.local
WC Version: 7.3.0
REST API Version: ✔ 7.3.0
WC Blocks Version: ✔ 8.9.1
Action Scheduler Version: ✔ 3.4.0
Log Directory Writable: ✔
WP Version: 6.1.1
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: nginx/1.16.0
PHP Version: 7.4.30
PHP Post Max Size: 1,000 MB
PHP Time Limit: 1200
PHP Max Input Vars: 4000
cURL Version: 7.84.0
(SecureTransport) OpenSSL/1.1.1q

SUHOSIN Installed: –
MySQL Version: 8.0.16
Max Upload Size: 300 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 7.3.0
WC Database Prefix: wp_
Total Database Size: 10.46MB
Database Data Size: 8.52MB
Database Index Size: 1.94MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.09MB + Index: 0.09MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.06MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 7.06MB + Index: 0.06MB + Engine InnoDB
wp_postmeta: Data: 0.20MB + Index: 0.19MB + Engine InnoDB
wp_posts: Data: 0.06MB + Index: 0.06MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_deposits_payment_plans: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_deposits_payment_plans_schedule: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_addresses: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_operational_data: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_orders: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_wc_orders_meta: Data: 0.05MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wcpv_commissions: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wcpv_per_product_shipping_rules: Data: 0.02MB + Index: 0.00MB + Engine InnoDB

### Post Type Counts ###

attachment: 26
custom_css: 1
event_ticket: 1
page: 9
post: 1
product: 24
product_variation: 7
revision: 4
shop_order: 27
shop_order_placehold: 25
wp_global_styles: 1

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (5) ###

Query Monitor: by John Blackbourn – 3.10.1
Debug Bar: by wordpressdotorg – 1.1.4
WooCommerce Deposits: by WooCommerce – 2.1.1
WooCommerce Stripe Gateway: by WooCommerce – 7.0.1
WooCommerce: by Automattic – 7.3.0-dev

### Inactive Plugins (7) ###

Braintree for WooCommerce Payment Gateway: by WooCommerce – 2.7.0 (update to version 2.9.0 is available)
WooCommerce Box Office: by WooCommerce – 1.1.40
WooCommerce GoCardless Gateway: by WooCommerce – 2.5.1
WooCommerce Pre-Orders: by WooCommerce – 1.7.3
WooCommerce Product Vendors: by WooCommerce – 2.1.66
WooCommerce SnapScan Gateway: by WooCommerce – 1.1.12
WooCommerce Subscriptions: by WooCommerce – 4.6.0

### Dropin Plugins (1) ###

db.php: Query Monitor Database Class (Drop-in)

### Must Use Plugins (1) ###

woo command to check compatibility.php: by  –

### Settings ###

API Enabled: –
Force SSL: –
Currency: ZAR (R)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
subscription (subscription)
variable (variable)
variable subscription (variable-subscription)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: –

### WC Pages ###

Shop base: #6 - /shop/
Cart: #7 - /cart/
Checkout: #8 - /checkout/
My account: #9 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Storefront
Version: 4.1.3 (update to version 4.2.0 is available)
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Admin ###

Enabled Features: activity-panels
analytics
coupons
customer-effort-score-tracks
experimental-products-task
experimental-import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
multichannel-marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: minified-js
new-product-management-experience
settings

Daily Cron: ✔ Next scheduled: 2022-12-10 04:19:12 +00:00
Options: ✔
Notes: 61
Onboarding: completed

### Action Scheduler ###

Complete: 21
Oldest: 2022-11-16 03:03:28 +0000
Newest: 2022-12-09 11:11:21 +0000

Pending: 1
Oldest: 2022-12-10 10:27:24 +0000
Newest: 2022-12-10 10:27:24 +0000

### Status report information ###

Generated at: 2022-12-09 11:30:21 +00:00
`

</details>

### Isolating the problem

- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by vedanshujain**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

In the posts table structure, when a post is deleted that has child orders, [WP would upshift all the child posts to the parent of deleted post](https://github.com/WordPress/WordPress/blob/master/wp-includes/post.php#L3420-L3427). We missed doing that for HPOS. This causes fatals when a parent post is deleted, but child posts are loaded, for example, by REST API or generally in the interface.

This PR fixes this issue by adding similar upshifts for child orders of deleted parent orders.

Closes #35909, #35850

- [x] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section). 
(Note: Dedicated testing is not required since it would be eventually covered in e2e tests in #36219)

### How to test the changes in this Pull Request:

(Unfortunately, there is no good way to simulate this using the interface, so we will use REST API to do it).
1. Create a new order using API request like so: `POST /wc/v3/orders`. Data can be set to `{}`.
2. Create another order, but specify the parent as the order created in step 1. For example: `POST /wc/v3/orders` with data `{ "parent_id": xxxx}` where xxxx is the ID of order created in step 1.
3. Delete the first order like so: `DELETE /wc/v3/orders/xxxx?force=true`
4. Get the order from step 2 again `GET /wc/v3/orders/{child_order_id`. Check that it's not fatal, and that the parent id is set to `0`.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

**Review by barryhughes — APPROVED** [review]
Nice, good to merge except there is a failing E2E check—which I am pretty sure is unrelated and so have triggered a re-run. Once we're green, though, let's merge :+1:

**Comment by vedanshujain** [comment]
@barryhughes , actually don't merge. seems like we are pretty close to enabling merge strategy, which if I can use, then I wouldn't have to resolve conflicts in #36219 eventually.

**Pull request merged**
